client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java

```diff
@@ -0,0 +1,508 @@
+import static io.trino.client.ClientStandardTypes.UUID;
+import static io.trino.client.ClientStandardTypes.VARCHAR;
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+public final class JsonDecodingUtils
+{
+    private JsonDecodingUtils() {}
+
+    static TypeDecoder[] createTypeDecoders(List<Column> columns)
+    {
+        requireNonNull(columns, "columns is null");
+        TypeDecoder[] decoders = new TypeDecoder[columns.size()];
+        int index = 0;
+        for (Column column : columns) {
+            decoders[index++] = createTypeDecoder(column.getTypeSignature());
+        }
+        return decoders;
+    }
+
+    interface TypeDecoder
+    {
+        Object decode(JsonParser parser)
+                throws IOException;
+    }
+
+    private static TypeDecoder createTypeDecoder(ClientTypeSignature signature)
+    {
+        switch (signature.getRawType()) {
+            case BIGINT:
+                return new BigIntegerDecoder();
+            case INTEGER:
+                return new IntegerDecoder();
+            case SMALLINT:
+                return new SmallintDecoder();
+            case TINYINT:
+                return new TinyintDecoder();
+            case DOUBLE:
+                return new DoubleDecoder();
+            case REAL:
+                return new RealDecoder();
+            case BOOLEAN:
+                return new BooleanDecoder();
+            case ARRAY:
+                return new ArrayDecoder(signature);
+            case MAP:
+                return new MapDecoder(signature);
+            case ROW:
+                return new RowDecoder(signature);
+            case VARCHAR:
+            case JSON:
+            case TIME:
+            case TIME_WITH_TIME_ZONE:
+            case TIMESTAMP:
+            case TIMESTAMP_WITH_TIME_ZONE:
+            case DATE:
+            case INTERVAL_YEAR_TO_MONTH:
+            case INTERVAL_DAY_TO_SECOND:
+            case IPADDRESS:
+            case UUID:
+            case DECIMAL:
+            case CHAR:
+            case GEOMETRY:
+            case SPHERICAL_GEOGRAPHY:
+                return new StringDecoder();
+            default:
+                return new Base64Decoder();
+        }
+    }
+
+    private static class BigIntegerDecoder
+            implements TypeDecoder
+    {
+        @Override
+        public Object decode(JsonParser parser)
+                throws IOException
+        {
+            switch (parser.nextToken()) {
+                case VALUE_NULL:
+                    return null;
+                case VALUE_NUMBER_INT:
+                    return parser.getLongValue();
+                case VALUE_STRING:
+                    return Long.parseLong(parser.getText());
+                default:
+                    return illegalToken(parser);
```

Comment:
The typical way is to do `throw illegalToken(parser)` so that the compiler knows this always throws

Reply:
makes sense